### PR TITLE
#16733: binary pow sfpu operation

### LINKED ISSUE #16733: ttnn.pow returns wrong values with exponent tensor of shape larger than [32,32]
**Describe the bug**
ttnn::pow seem to have issues when exponent tensor shape is larger than a single tile

**To Reproduce**

Running with the latest tt-metal. Python code with shape higher than 32,32 (any shape higher than it  does not work):
```py

import torch
import ttnn
import numpy as np
def compare_tensors_with_threshold(tensor1, tensor2, threshold=1e-5):
    # Get absolute differences
    diff = torch.abs(tensor1 - tensor2)
    # Find positions where difference exceeds threshold
    diff_mask = diff > threshold
    diff_positions = torch.nonzero(diff_mask)
    # Get values
    values1 = tensor1[diff_mask]
    values2 = tensor2[diff_mask]
    # Print significant differences
    n = 5  # Stop after 5 items
    for i, (pos, v1, v2) in enumerate(zip(diff_positions, values1, values2), start=1):
        if i > n:
            break
        print(f"#{i} Position {tuple(pos.tolist())}: {v1:.6f} vs {v2:.6f}, diff = {abs(v1-v2):.6f}")
    return diff_mask.any()
torch.manual_seed(0)
device_id = 0
device = ttnn.open_device(device_id=device_id)
ttnn.enable_program_cache(device)
torch_a = torch.randn((32, 64), dtype=torch.bfloat16)
torch_b = torch.randn((32, 64), dtype=torch.bfloat16)
a = ttnn.from_torch(torch_a, layout=ttnn.TILE_LAYOUT, device=device)
b = ttnn.from_torch(torch_b, layout=ttnn.TILE_LAYOUT, device=device)
output = ttnn.pow(a, b);
torch_output = torch.pow(torch_a,torch_b)
output = ttnn.to_layout(output, ttnn.ROW_MAJOR_LAYOUT)
print("Printing ttnn tensor")
print(f"shape: {output.shape}")
#print(f"chunk of a tensor:\n{output[:1, :32]}")
output_torch = output.cpu().to_torch()
rtol = 1.
equal = torch.allclose(torch_output, output_torch, rtol) #print(torch.allclose(torch_output, output_torch, rtol))
threshold = 20. #in np.arange(0.5, 100., 0.5):
    # do something
if not equal:
    if compare_tensors_with_threshold(torch_output, output_torch, threshold):
        print("Failed with rtol: ", rtol, "trshld: ", threshold)
    else:
        print("Passed with rtol: ", rtol, "trshld: ", threshold)
else:
    print("Passed with rtol: ", rtol, "trshld: ", threshold)
ttnn.close_device(device)
```
 Here is the result:
```console
                  Metal | INFO     | Closing device 0
                  Metal | INFO     | Disabling and clearing program cache on device 0
                 Device | INFO     | Closing user mode device drivers
(python_env) user@yyz-cust1:~/ox/gpu.shared$ python ./ox-src/examples/oxtosagpu/python/oxttnnops.py
2025-01-14 20:40:49.966 | DEBUG    | ttnn:<module>:82 - Initial ttnn.CONFIG:
Config{cache_path=/home/user/.cache/ttnn,model_cache_path=/home/user/.cache/ttnn/models,tmp_dir=/tmp/ttnn,enable_model_cache=false,enable_fast_runtime_mode=true,throw_exception_on_fallback=false,enable_logging=false,enable_graph_report=false,enable_detailed_buffer_report=false,enable_detailed_tensor_report=false,enable_comparison_mode=false,comparison_mode_pcc=0.9999,root_report_path=generated/ttnn/reports,report_name=std::nullopt,std::nullopt}
2025-01-14 20:40:50.031 | WARNING  |                  Device | INFO     | Opening user mode device driver
  Detecting chips (found 1)
2025-01-14 20:40:50.069 | INFO     | SiliconDriver   - Opened PCI device 0; KMD version: 1.29.0, IOMMU: disabled
2025-01-14 20:40:50.069 | INFO     | SiliconDriver   - Detected PCI devices: [0]
2025-01-14 20:40:50.069 | INFO     | SiliconDriver   - Using local chip ids: {0} and remote chip ids {}
2025-01-14 20:40:50.089 | INFO     | SiliconDriver   - Software version 6.0.0, Ethernet FW version 6.9.0 (Device 0)
                  Metal | INFO     | Initializing device 0. Program cache is NOT enabled
                  Metal | INFO     | AI CLK for device 0 is:   1000 MHz
                  Metal | INFO     | Enabling program cache on device 0
Printing ttnn tensor
shape: ttnn.Shape([32, 64])
#1 Position (0, 40): 1.000000 vs -1247611445842297308296773632.000000, diff = 1247611445842297308296773632.000000
#2 Position (0, 41): 0.890625 vs -1702167554017397877986295808.000000, diff = 1702167554017397877986295808.000000
#3 Position (0, 56): 0.863281 vs -1876252872041904479143985152.000000, diff = 1876252872041904479143985152.000000
#4 Position (1, 34): 0.211914 vs -507748844238144253376593920.000000, diff = 507748844238144253376593920.000000
#5 Position (1, 35): 0.644531 vs -4564903894864839763690520576.000000, diff = 4564903894864839763690520576.000000
Failed with rtol:  1.0 trshld:  20.0
                  Metal | INFO     | Closing device 0
                  Metal | INFO     | Disabling and clearing program cache on device 0
                 Device | INFO     | Closing user mode device drivers (edited) 
```

I though the difference if with data type but here is the result with float32:
```console
Printing ttnn tensor
shape: ttnn.Shape([32, 64])
#1 Position (0, 35): 0.073002 vs -131008.000000, diff = 131008.070312
#2 Position (0, 36): 0.437620 vs -131008.000000, diff = 131008.437500
#3 Position (0, 37): 0.149081 vs -131008.000000, diff = 131008.148438
#4 Position (0, 41): 0.874949 vs -131008.000000, diff = 131008.875000
#5 Position (0, 43): 0.048576 vs -131008.000000, diff = 131008.046875
Failed with rtol:  1.0 trshld:  20.0
```

### PR DESCRIPTION
**Pull request by KalaivaniMCW**
### Ticket
Link to Github Issue #16733 

### Problem description
Incorrect result for certain values in `ttnn.pow`
required LLK side fixes merged in #17267 

### What's changed
In binary device operation, based on `dtype` we choose the FPU (bfloat16) or SFPU (float32) operation for compute. Binary Pow is an exception here, regardless of dtype the operation runs on SFPU.
For all SFPU ops - UnpackToDestMode is set to `UnpackToDestFp32`, hence adding another check to set this only if input dtype's not bfloat16.
**Observation:**
As Radomir confirmed with the HW team, In bfloat16 dtype, NaN values should become Inf when they get packed out. This is a HW limitation. For float32, NaN values are unaffected.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/13179137515
- [x] Blackhole Post commit (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/13182941436
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by VirdhatchaniKN — APPROVED** [review]

**Review by rdjogoTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_pow.py [hunk 1]
```diff
@@ -182,3 +182,73 @@ def test_binary_sfpu_pow_neg(
...
+    x_tt = ttnn.from_torch(x_torch, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
+    y_tt = ttnn.from_torch(y_torch, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
+    z_tt_pow = ttnn.pow(x_tt, y_tt)
+    tt_out = ttnn.to_torch(z_tt_pow)
+    # Due to HW limitations for bfloat16 dtype, NaN value gets packed as inf.
+    # z_tt_pow ttnn.Tensor([[ 0.99609,  0.67969,  ...,  0.89844,      inf]])
+    # z_torch tensor([[1.0000, 0.6758, 0.8906,    nan]], dtype=torch.bfloat16)
+
+    status = ttnn.pearson_correlation_coefficient(z_torch, tt_out) >= 0.99
+    assert status
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/operations/eltwise/test_pow.py:230]: Was it intentional to `skip_for_grayskull` on bf16? If bf16 is supposed to run on Grayskull, then do this:
```suggestion
@pytest.mark.parametrize(
    "dtype",
    [
        "float32",
        "bfloat16",
    ],
)
def test_binary_pow(device, dtype):
    if dtype == "float32" and is_grayskull():
        pytest.skip("Unsupported dtype for Grayskull")

    torch_dtype = getattr(torch, dtype)
    ttnn_dtype = getattr(ttnn, dtype)
    x_torch = torch.tensor([[0.98828125, 0.47851562, 1.1875, -1.59375]], dtype=torch_dtype)
    y_torch = torch.tensor([[0.0751953125, 0.53125, -0.6640625, 0.1533203125]], dtype=torch_dtype)
    z_torch = torch.pow(x_torch, y_torch)
    x_tt = ttnn.from_torch(x_torch, dtype=ttnn_dtype, layout=ttnn.TILE_LAYOUT, device=device)
    y_tt = ttnn.from_torch(y_torch, dtype=ttnn_dtype, layout=ttnn.TILE_LAYOUT, device=device)
    z_tt_pow = ttnn.pow(x_tt, y_tt)
    tt_out = ttnn.to_torch(z_tt_pow)

    status = ttnn.pearson_correlation_coefficient(z_torch, tt_out) >= 0.99
    assert status
```
otherwise just use `@skip_for_grayskull` decorator to skip for both dtypes.

File: ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp [hunk 2]
```diff
@@ -541,7 +543,9 @@ BinaryNgDeviceOperation::ProgramFactory::cached_program_t BinaryNgDeviceOperatio
     uint32_t src1interim_cb_index = tt::CBIndex::c_4;
 
     std::vector<UnpackToDestMode> unpack_to_dest_mode(NUM_CIRCULAR_BUFFERS, UnpackToDestMode::Default);
-    if (is_sfpu_op) {
+
+    bool is_bf16 = (a_dtype == DataType::BFLOAT16 && b_dtype == DataType::BFLOAT16);
+    if (is_sfpu_op && !is_bf16) {
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp:548]: Shouldn't the `unpack_to_dest_mode` be checked for `src0` and `src1` independently? Add some test cases in Python for mixed float32 / bfloat16 operands to make sure this is correct.

**Review by patrickroberts — CHANGES_REQUESTED** [review]

**Review by mcw-anasuya — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_pow.py (hunk 1 above)
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_pow.py:230]: The sfpu ops are available only for wormhole and blackhole 
https://github.com/tenstorrent/tt-metal/pull/15782/files

File: ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp (hunk 2 above)
Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp:548]: Mixed datatypes would cause a lower PCC due to precision loss and also the inf/nan limitation with bfloat16

File: tests/ttnn/unit_tests/operations/eltwise/test_pow.py (hunk 1 above)
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_pow.py]: I have updated the tests with both dtypes

**Review by patrickroberts — APPROVED** [review]
